service/src/main/java/bio/terra/drshub/services/AuthService.java

```diff
@@ -0,0 +1,243 @@
+package bio.terra.drshub.services;
+
+import bio.terra.bond.model.SaKeyObject;
+import bio.terra.common.iam.BearerToken;
+import bio.terra.drshub.DrsHubException;
+import bio.terra.drshub.config.DrsProvider;
+import bio.terra.drshub.models.AccessUrlAuthEnum;
+import bio.terra.drshub.models.DrsHubAuthorization;
+import com.google.common.annotations.VisibleForTesting;
+import io.github.ga4gh.drs.model.AccessMethod;
+import io.github.ga4gh.drs.model.Authorizations;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections4.map.PassiveExpiringMap;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpStatusCodeException;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.util.UriComponents;
+
+@Service
+@Slf4j
+public class AuthService {
+
+  private final BondApiFactory bondApiFactory;
+  private final DrsApiFactory drsApiFactory;
+  private final ExternalCredsApiFactory externalCredsApiFactory;
+
+  // To avoid absolutely hammering the ECM API during large batch analyses,
+  // cache the passport for a given bearer token for just a little bit.
+  // This also keeps DRSHub from calling ECM twice for the same request
+  // if the object info endpoint needs passport auth as well as the object access url endpoint.
+  private final Map<String, Optional<List<String>>> passportCache =
+      Collections.synchronizedMap(new PassiveExpiringMap<>(1, TimeUnit.MINUTES));
```

Comment:
this seems really useful! Should we get a security review for this since it does mean access could be out of date? It seems pretty minimal risk (being ~1 minute out of sync) but just in case that's important for Compliance.

Reply:
I'm pretty confident that this won't be an issue. Other apps cache PET accounts for a small amount of time without issue as an optimization.  Here's an example in TDR: https://github.com/DataBiosphere/jade-data-repo/blob/96deee3ac6964795d07a6106b8e0e8618da0be9b/src/test/java/bio/terra/common/auth/AuthService.java#L51-L58

Reply:
We could, but in the spirit of minimal changes, we should do that in a separate PR.